pygen/pygen_src/riscv_directed_instr_lib.py

```diff
@@ -161,3 +163,134 @@ def post_randomize(self):
+        for i in range(len(self.push_stack_instr)):
+            self.push_stack_instr[i] = riscv_instr()
+        self.push_stack_instr[0] = \
+            riscv_instr_ins.get_instr(riscv_instr_name_t.ADDI.name)
+        with self.push_stack_instr[0].randomize_with() as it:
+            self.push_stack_instr[0].rd == cfg.sp
+            self.push_stack_instr[0].rs1 == cfg.sp
+            self.push_stack_instr[0].imm == cfg.stack_len + 1
+
+        self.push_stack_instr[0].imm_str = '-{}'.format(self.stack_len)
+        for i in range(len(self.saved_regs)):
+            if rcs.XLEN == 32:
+                self.push_stack_instr[i + 1] = riscv_instr_ins.get_instr(riscv_instr_name_t.SW.name)
+                with self.push_stack_instr[i + 1].randomize_with() as it:
+                    self.push_stack_instr[i + 1].rs2 == self.saved_regs[i]
+                    self.push_stack_instr[i + 1].rs1 == cfg.sp
+                    self.push_stack_instr[i + 1].imm == 4 * (i + 1)
+            else:
+                self.push_stack_instr[i + 1] = riscv_instr_ins.get_instr(riscv_instr_name_t.SD.name)
+                with self.push_stack_instr[i + 1].randomize_with() as it:
+                    self.push_stack_instr[i + 1].rs2 == self.saved_regs[i]
+                    self.push_stack_instr[i + 1].rs1 == cfg.sp
+                    self.push_stack_instr[i + 1].imm == 8 * (i + 1)
+
+            self.push_stack_instr[i + 1].process_load_store = 0
+        if allow_branch:
+            # TODO `DV_CHECK_STD_RANDOMIZE_FATAL(enable_branch)
+            pass
+        else:
+            self.enable_branch = 0
+
+        if self.enable_branch:
+            self.branch_instr = \
+                riscv_instr_ins.get_rand_instr(include_category=[riscv_instr_name_t.BRANCH.name])
+            # TODO `DV_CHECK_STD_RANDOMIZE_FATAL(branch_instr)
+            self.branch_instr.imm_str = self.push_start_label
+            self.branch_instr.brach_assigned = 1
+            self.push_stack_instr[0].label = self.push_start_label
+            self.push_stack_instr[0].has_label = 1
+            self.branch_instr.extend(self.push_stack_instr)
+        self.mix_instr_stream(self.push_stack_instr)
+        for i in range(len(self.instr_list)):
+            self.instr_list[i].atomic = 1
+            if self.instr_list[i].label == '':
+                self.instr_list[i].has_label = 0
+
+
+# Pop stack instruction stream
+class riscv_pop_stack_instr(riscv_rand_instr_stream):
+
+    def __init__(self):
+        super().__init__()
+        self.stack_len = 0
+        self.num_of_reg_to_save = 0
+        self.num_of_redudant_instr = 0
+        self.pop_stack_instr = []
+        self.saved_regs = []
+
+    def init(self):
+        self.reserved_rd = [cfg.ra]
+        self.num_of_reg_to_save = len(self.saved_regs)
+        if self.num_of_reg_to_save * 4 > self.stack_len:
+            logging.critical('stack len [%0d] is not enough to store %d regs',
```

Comment:
nit: in the `init()` function of `riscv_push_stack_instr` a `logging.error` call is used instead, maybe the  same logging severity should  be used in both contexts?

Reply:
Yeah, makes sense.
Fixed now!